test/libdnf/impl_ptr/test_impl_ptr.cpp

```diff
@@ -0,0 +1,164 @@
+        CPPUNIT_ASSERT_EQUAL(2, CTest::get_instance_counter());
+        CPPUNIT_ASSERT_EQUAL(10, object3->get_a());
+        CPPUNIT_ASSERT(nullptr == object2.get());
+
+        // Tests the copy assignment operator. The source object remains unchanged.
+        // The value of the `CTest` instance in the source is copied to the CTest instance in the destination.
+        libdnf::ImplPtr<CTest> object4(new CTest);
+        CPPUNIT_ASSERT_EQUAL(3, CTest::get_instance_counter());
+        CPPUNIT_ASSERT_EQUAL(0, object4->get_a());
+        object4 = object1;
+        CPPUNIT_ASSERT_EQUAL(3, CTest::get_instance_counter());
+        CPPUNIT_ASSERT_EQUAL(10, object4->get_a());
+        CPPUNIT_ASSERT(nullptr != object1.get());
+        CPPUNIT_ASSERT_EQUAL(10, object1->get_a());
+
+        // Tests the copy assignment operator - self-assignment.
+        // The object must remain unchanged.
+        object4 = object4;
+        CPPUNIT_ASSERT_EQUAL(3, CTest::get_instance_counter());
+        CPPUNIT_ASSERT_EQUAL(10, object4->get_a());
+
+        // Tests the move assignment operator.
+        // `CTest` instance ownership is transferred from the source object to the destination.
+        // The old `CTest` instance in the destination is destroyed.
+        // `p_impl` in the original object becomes nullptr.
+        libdnf::ImplPtr<CTest> object5(new CTest);
+        CPPUNIT_ASSERT_EQUAL(4, CTest::get_instance_counter());
+        CPPUNIT_ASSERT_EQUAL(0, object5->get_a());
+        object5 = std::move(object4);
+        CPPUNIT_ASSERT_EQUAL(3, CTest::get_instance_counter());
+        CPPUNIT_ASSERT_EQUAL(10, object5->get_a());
+        CPPUNIT_ASSERT(nullptr == object4.get());
+
+        // Tests the move assignment operator - self-assignment.
+        // The object must remain unchanged.
+        object5 = std::move(object5);
+        CPPUNIT_ASSERT_EQUAL(3, CTest::get_instance_counter());
+        CPPUNIT_ASSERT_EQUAL(10, object5->get_a());
+
+        // Tests the copy assignment to empty (moved from) object. The source object remains unchanged.
+        // A new `CTest` instance is created.
+        CPPUNIT_ASSERT(nullptr == object2.get());
+        object2 = object5;
+        CPPUNIT_ASSERT_EQUAL(4, CTest::get_instance_counter());
+        CPPUNIT_ASSERT_EQUAL(10, object2->get_a());
+        CPPUNIT_ASSERT_EQUAL(10, object5->get_a());
+
+        // Tests the move assignment to empty (moved from) object.
+        // `CTest` instance ownership is transferred from the source object to the destination.
+        // `p_impl` in the original object becomes nullptr.
+        CPPUNIT_ASSERT(nullptr == object4.get());
+        object4 = std::move(object5);
+        CPPUNIT_ASSERT_EQUAL(4, CTest::get_instance_counter());
+        CPPUNIT_ASSERT_EQUAL(10, object4->get_a());
+        CPPUNIT_ASSERT(nullptr == object5.get());
+
+        // Tests the `operator*()`.
+        CPPUNIT_ASSERT_EQUAL((*object4).get_a(), object4->get_a());
+    }
```

Comment:
Could the above once again be split into multiple smaller tests?

Otherwise looks good.

Reply:
Ok. Done.

Reply:
Thanks!